nfs.py

```diff
@@ -4,13 +4,16 @@
 from logger import logger
 from typing import Optional
 import sys
+import threading
 
 
 """
 NFS is needed in many cases to network mount the folder that contains
 ISO files such that Red Fish Virtual Media managers can load the image.
 """
 
+_lock = threading.Lock()
```

Comment:
Shouldn't the lock be tied to a host?

Reply:
commit message:

    For the same reason is the `_lock` at module level, instead in nfs.NFS.
    The operations we guard are those on the system (`systemctl restart`),
    and we don't want that two nfs.NFS instances could interfere.

Reply:
The commit message mentions:  `However, there is no downside in adding an (somewhat far-reaching) additional lock while calling the NFS API.`  Isn't locking on the host that you run nfs on (i.e. the host that you get in the constructor) more precise and aligned with what readers of the code would expect?

Reply:
This issue breaks deployment today, but 185 did not materialize and no solution is immanent. 

Also, it relies on using the same Host variable. Which coincidentally is probably the case today, given the (questionable?) internalizing in `Host.__new__()`. The code should not rely on that. Also, the only reason why `Host.run()` will need locking, is because paramiko is not thread safe and the instance has mutable state, and host instances are shared. In general, there is no issue with running two processes in parallel on the same host.

Also, the locking around `Host.run()` only guards one command, while the lock here is around the entire operation. For example, there is still a race in
```
    def _add(self, dir_name: str) -> None:
        contents = self._host.read_file("/etc/exports")
        self._host.write("/etc/exports", f"{contents}\n{dir_name}")
```
if you don't lock the entire operation.


There is no downside, because the `NFS.host_file()` operation tends to be fast (enough), so having another thread waiting is not an issue.


The only limitation of the lock is that the entire code is not robust against two processes fighting over /etc/exports file. But that is a common problem, and the solution is: don't have two processes running that fight over the same files and system resources. Which in practice holds well enough. Somewhat unusual, this lock does not protect data of the python program itself, it exists to perform the operations on the system without interference from other threads.


There is no conflict with a followup to #185. That one will still be needed, for other reasons.

Reply:
Can we have a lock in nfs that is unique per host?